### PR TITLE
feat: @keyframes now support <string> type AND will convert it into ident if possible...

### PR DESCRIPTION
**Pull request by CGQAQ**
... Plus `animation` and `animation-name` mangling as well

fixes: #2555

### CODE REVIEW

File: internal/css_parser/css_parser_test.go [hunk 1]
```diff
@@ -1013,6 +1013,23 @@ func TestLegalComment(t *testing.T) {
...
 	expectPrinted(t, "@keyframes name{}", "@keyframes name {\n}\n")
+	// string name valid to convert into ident
+	expectPrinted(t, `@keyframes "name"{}`, "@keyframes \"name\" {\n}\n")
+	expectPrintedMinify(t, `@keyframes "name"{}`, "@keyframes \"name\"{}")
+	expectPrintedMangleMinify(t, `@keyframes "name"{}`, "@keyframes name{}")
+
+	// string name valid to convert into ident but need escape
+	expectPrinted(t, `@keyframes "foo bar"{}`, "@keyframes \"foo bar\" {\n}\n")
+	expectPrintedMinify(t, `@keyframes "foo bar"{}`, "@keyframes \"foo bar\"{}")
+	expectPrintedMangleMinify(t, `@keyframes "foo bar"{}`, "@keyframes foo\\ bar{}")
```
Comment (yisibl) [inline internal/css_parser/css_parser_test.go:1024]: Add test `@keyframes "none" {}`

Comment (yisibl) [reply internal/css_parser/css_parser_test.go:1024]: Because `none` is not included in the definition of `cssWideAndReservedKeywords`.

https://github.com/evanw/esbuild/blob/4a6f64cd81c996aa2868f467ae34e9af46405861/internal/css_parser/css_decls_font_family.go#L24-L34

File: internal/css_parser/css_parser.go [hunk 2]
```diff
@@ -828,11 +828,21 @@ abortRuleParser:
...
+			isStringName = false
+		} else if p.peek(css_lexer.TString) {
+			name = strings.ToLower(p.decoded())
+			isStringName = !p.options.MinifySyntax || cssWideAndReservedKeywords[name] || name == "none"
 			p.advance()
-		} else if !p.expect(css_lexer.TIdent) && !p.eat(css_lexer.TString) && !p.peek(css_lexer.TOpenBrace) {
+		} else if !(p.expect(css_lexer.TIdent) && p.expect(css_lexer.TString)) && !p.peek(css_lexer.TOpenBrace) {
 			// Consider string names a syntax error even though they are allowed by
 			// the specification and they work in Firefox because they do not work in
 			// Chrome or Safari.
```
Comment (evanw) [inline internal/css_parser/css_parser.go:848]: Not sure if you saw this comment or not but lack of support for this was deliberate. This is currently a Firefox-only feature, which I don't think makes sense to support natively in esbuild at this time.

I agree that the bug that changes the contents of `@keyframes` rules with unsupported syntax should still be fixed. These should be passed through unmodified. I will fix that part.

Comment (CGQAQ) [reply internal/css_parser/css_parser.go:848]: I currently working on blink, and the webkit one is already merged

ref: https://github.com/WebKit/WebKit/pull/4234
      https://chromium-review.googlesource.com/c/chromium/src/+/3865188

**Comment by CGQAQ** [comment]
I saw you submitted a [commit](https://github.com/evanw/esbuild/commit/c9b24d829cfb8a08d4ade8d7d78bee568f163270) to fix #2555 which makes it an unrecognized CSS rule,  I think it need to be reverted to land my PR land to conform the [CSS standard](https://www.w3.org/TR/css-animations-1/#typedef-keyframes-name
)

**Comment by evanw** [comment]
> I think it need to be reverted to land my PR land to conform the [CSS standard](https://www.w3.org/TR/css-animations-1/#typedef-keyframes-name)

It should already confirm to the CSS standard after the latest release. The syntax should now be passed through unmodified, so you should now be able to use this syntax with esbuild's latest release. Do you have some CSS code for which that isn't true?

File: internal/css_parser/css_parser.go (hunk 2 above)
Comment (yisibl) [reply internal/css_parser/css_parser.go:848]: This is not a Firefox-only syntax, the prefixed `@-webkit-keyframes / @-moz-keyframes` always support `<string>`, and there are [UseCounter](https://codereview.chromium.org/651123003/) in Chrome that say: `The usage counter for strings in prefixed case is too high`.

**Comment by evanw** [comment]
I believe the commit you linked to in your previous comment should already implement what this PR is trying to do. One difference is that the string is always converted to an identifier, but this is because Chrome has refused to implement support for string names (at least as far as I can tell). Leaving the string name as-is would result in broken code in all Chromium-based browsers.
